File: ios/bitchat/bitchat/V2V/Models/TransportChannel.swift
import Foundation

public enum TransportChannel {
    case bleMesh
    case firebaseCloud
    case nostrRelay
}

public enum MessageProcessResult {
    case processed(messageId: String, channel: TransportChannel, latencyMs: Int64)
    case duplicate(messageId: String, originalChannel: TransportChannel, duplicateChannel: TransportChannel)
    case invalid(messageId: String?, reason: String)
}

public struct IncomingPacket: Equatable {
    public let data: Data
    public let channel: TransportChannel
    public let receivedAtMs: Int64
    public let metadata: [String: String]

    public init(
        data: Data,
        channel: TransportChannel,
        receivedAtMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        metadata: [String: String] = [:]
    ) {
        self.data = data
        self.channel = channel
        self.receivedAtMs = receivedAtMs
        self.metadata = metadata
    }

    public static func == (lhs: IncomingPacket, rhs: IncomingPacket) -> Bool {
        return lhs.data == rhs.data && lhs.channel == rhs.channel
    }
}

extension TransportChannel: Equatable {}

public protocol MessageTransport: AnyObject {
    var channel: TransportChannel { get }
    var isAvailable: Bool { get }
    func start() async
    func stop() async
    func send(packet: Data, targetGeohash: String?) async throws

    var onIncoming: ((IncomingPacket) -> Void)? { get set }
}

File: app/src/main/java/com/cybersiren/android/online/TransportChannel.kt
package com.cybersiren.android.online

enum class TransportChannel {

    BLE_MESH,

    FIREBASE_CLOUD,

    NOSTR_RELAY
}

sealed class MessageProcessResult {

    data class Processed(
        val messageId: String,
        val channel: TransportChannel,
        val latencyMs: Long
    ) : MessageProcessResult()

    data class Duplicate(
        val messageId: String,
        val originalChannel: TransportChannel,
        val duplicateChannel: TransportChannel
    ) : MessageProcessResult()

    data class Invalid(
        val messageId: String?,
        val reason: String
    ) : MessageProcessResult()
}

interface MessageTransport {

    val channel: TransportChannel

    val isAvailable: Boolean

    suspend fun send(packet: ByteArray, targetGeohash: String? = null)

    fun observeIncoming(): kotlinx.coroutines.flow.Flow<IncomingPacket>

    suspend fun start()

    suspend fun stop()
}

data class IncomingPacket(
    val data: ByteArray,
    val channel: TransportChannel,
    val receivedAtMs: Long = System.currentTimeMillis(),
    val metadata: Map<String, String> = emptyMap()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as IncomingPacket
        return data.contentEquals(other.data) && channel == other.channel
    }

    override fun hashCode(): Int {
        var result = data.contentHashCode()
        result = 31 * result + channel.hashCode()
        return result
    }
}
